# Evaluating The Performance of Various Bayesian Models

## Objective
Leveraging different Regression Models in a range of contexts

### Set-Up

The module will compare the performance of six Bayesian regression models: Linear regression, Ridge and Lasso Regression, ICR and PCR, Variational Inference and Robust Regression
It will compare their performance either using RMSE or MSE depending on if the true regression coefficients are known. 

Given the large number of features in many datasets nowadays and how time consuming sampling from a posterior can be, the code will incorporate a time out feature that will first evaluate the run time on a short chain with a very small number of draws in order to evaluate if sampling from the full posterior is infeasible. If that is the case, the module will skip that model and not evaluate its performance for that set-up.

### Implementation

#### Initial Stage

Using synthetic data, the first set-up will involve 3 predictors and an intercept term, the true coefficients will be obtained from a log space with 4 points between 1 and 2 and the design matrix will be fixed to reduce experimental randomness.

#### Experiments

The performance of the different regression models will be tested by playing around with a range of features, including the number of parameters, the sample size, the magnitude of the coefficients (both the intercept and the other beta coefficients). We will also change the design matrix by using highly sparse data which will only have three real parameters and increasingly more irrelevant features.

#### Real Data

The performance of the module will be evaluated with real data using three datasets. 

1. A housing dataset
2. Soccer player's market value
3. Cancer survival


In [1]:
import fourtenmodule
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import pandas as pd
"""from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.feature_selection import VarianceThreshold
"""

'from sklearn.preprocessing import StandardScaler, OneHotEncoder\nfrom sklearn.compose import ColumnTransformer\nfrom sklearn.pipeline import Pipeline\nfrom sklearn.impute import SimpleImputer\nfrom sklearn.feature_selection import VarianceThreshold\n'

## Synthetic Data

### Initial Set-Up

In [10]:
np.random.seed(410)
n=50
p = 5
# Create alternating columns of [0, 1] and [1, 0] in 3 columns to reduce experimental randomness
X = np.ones((n, p))
for i in range(1, X.shape[1], 2):
    X[i::2, i] = 0
    X[i+1::2, i] = 1
# Create true_beta based on logspace
true_beta = np.logspace(0, 1, p, base=2)
sigma_true = 1
print(true_beta)

# Generate y with added random noise
y = np.dot(X, true_beta) + stats.norm(0,sigma_true).rvs(n)

data_matrix = np.column_stack((X, y.reshape(-1, 1)))

feature_columns = [f"feature_{i+1}" for i in range(p)]
column_names = feature_columns + ["response"]
df = pd.DataFrame(data_matrix, columns=column_names)

# Verify the DataFrame
print("\nDataFrame with named columns:")
print(df.head())
print("\nDataFrame shape:", df.shape)
print("Column names:", df.columns.tolist())


fourtenmodule.run_and_plot_models(df, "response", true_beta, n_components=2)

Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


[1.         1.18920712 1.41421356 1.68179283 2.        ]

DataFrame with named columns:
   feature_1  feature_2  feature_3  feature_4  feature_5  response
0        1.0        1.0        1.0        1.0        1.0  7.589102
1        1.0        0.0        1.0        1.0        1.0  6.914526
2        1.0        1.0        1.0        1.0        1.0  8.235792
3        1.0        0.0        1.0        0.0        1.0  4.369556
4        1.0        1.0        1.0        1.0        1.0  7.296018

DataFrame shape: (50, 6)
Column names: ['feature_1', 'feature_2', 'feature_3', 'feature_4', 'feature_5', 'response']
start
    feature_1  feature_2  feature_3  feature_4  feature_5
0         0.0        1.0        0.0   0.960769        0.0
1         0.0       -1.0        0.0   0.960769        0.0
2         0.0        1.0        0.0   0.960769        0.0
3         0.0       -1.0        0.0  -1.040833        0.0
4         0.0        1.0        0.0   0.960769        0.0
5         0.0       -1.0        0.0  -

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.79s

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.99s

Testing Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 1 seconds.
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.62s

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 0.86s

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 166.17
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.01s

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.79s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.58s

=== Running full analysis on passing models ===

Running Bayesian Linear Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 12 seconds.
There were 143 divergences after tuning. Increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 3 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 3 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 166.16
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 6.0752     β0:9.715  β1:0.292  β2:7.496  β3:1.174  β4:-9.748
Bayesian Linear Regression     6.0879     β0:1.000  β1:0.682  β2:1.420  β3:1.185  β4:1.996
Bayesian Robust Regression     6.0890     β0:1.006  β1:0.683  β2:1.410  β3:1.200  β4:1.995
Bayesian Variational Inference 6.0900     β0:1.028  β1:0.712  β2:1.382  β3:1.186  β4:2.045
Bayesian PCR                   6.0905     β0:-0.000  β1:0.497  β2:0.000  β3:0.560  β4:-0.000
Bayesian Ridge Regression      6.1002     β0:0.983  β1:0.760  β2:1.439  β3:1.269  β4:2.019
Bayesian ICR                   6.1085     β0:0.000  β1:0.160  β2:0.000  β3:0.674  β4:0.000


{'passing_models': {'Bayesian Linear Regression': <function fourtenmodule.bayesian_regression_mcmc(X, y, true_beta=None, draws=4000)>,
  'Bayesian Ridge Regression': <function fourtenmodule.bayesian_ridge_regression(X, y, true_beta=None, draws=4000)>,
  'Bayesian Lasso': <function fourtenmodule.bayesian_lasso(X, y, true_beta=None, n_folds=5, draws=4000)>,
  'Bayesian Robust Regression': <function fourtenmodule.bayesian_robust_regression(X, y, true_beta=None, draws=4000)>,
  'Bayesian Variational Inference': <function fourtenmodule.bayesian_regression_vi(X, y, true_beta=None)>,
  'Bayesian PCR': <function fourtenmodule.bayesian_pcr(X, y, true_beta=None, n_components=None, draws=4000)>,
  'Bayesian ICR': <function fourtenmodule.bayesian_icr(X, y, true_beta=None, n_components=None, draws=4000)>},
 'results': [{'Model': 'Bayesian Linear Regression',
   'RMSE': np.float64(6.087879377447663),
   'Beta Estimates': array([0.99953267, 0.68166317, 1.42005552, 1.18494576, 1.99612518])},
  {'Model

The true beta is given by [1, 1.18920712, 1.41421356, 1.68179283, 2]. The best model by RMSE appears to be the Bayesian Lasso Regression, with Linear and Robust regression producing similar scores. On the other hand, ICR and PCR produce significantly worse RMSE values due to the small number of predictors.


## Changing Parameters

### Sample Size

In [2]:
import matplotlib.pyplot as plt
import pandas as pd

# Define sample sizes to test
sample_sizes = [50, 100, 500, 1000, 1500]
model_names = [
    "Bayesian Linear Regression",
    "Bayesian Ridge Regression",
    "Bayesian Lasso", 
    "Bayesian Robust Regression",
    "Bayesian Variational Inference",
    "Bayesian PCR",
    "Bayesian ICR"
]

# Initialize results storage
results_data = []

for n in sample_sizes:
    print(f"\n=== Running models for sample size: {n} ===")
    
    # Get results for current sample size
    current_results = fourtenmodule.run_models_and_evaluate(n=n)
    
    # Verify results structure exists
    if 'results' not in current_results:
        print(f"Warning: No 'results' key found for n={n}")
        continue
        
    # Store results
    for result in current_results['results']:
        if result['Model'] in model_names:
            results_data.append({
                'Sample Size': n,
                'Model': result['Model'],
                'RMSE': result['RMSE']
            })

# Convert to DataFrame
results_df = pd.DataFrame(results_data)

# Pivot for plotting if we have data
if not results_df.empty:
    plot_data = results_df.pivot(index='Sample Size', columns='Model', values='RMSE')
    
    # Plot RMSE as a function of sample size
    plt.figure(figsize=(10, 6))
    for model in model_names:
        if model in plot_data.columns:
            plt.plot(plot_data.index, plot_data[model], marker='o', label=model)
            
    plt.xlabel("Sample Size")
    plt.ylabel("RMSE")
    plt.title("Model Performance vs Sample Size")
    plt.legend()
    plt.grid(True)
    plt.show()
else:
    print("No valid results to plot")

# Show complete results
print("\nComplete Results:")
print(results_df)


=== Running models for sample size: 50 ===
heeleloeo

=== Running with parameters: {'high_corr': False, 'n_components': 2} ===
start
    feature_0  feature_1  feature_2
0        -1.0        0.0   0.960769
1         1.0        0.0   0.960769
2        -1.0        0.0  -1.040833
3         1.0        0.0   0.960769
4        -1.0        0.0  -1.040833
5         1.0        0.0   0.960769
6        -1.0        0.0  -1.040833
7         1.0        0.0   0.960769
8        -1.0        0.0  -1.040833
9         1.0        0.0   0.960769
10       -1.0        0.0  -1.040833
11        1.0        0.0   0.960769
12       -1.0        0.0  -1.040833
13        1.0        0.0   0.960769
14       -1.0        0.0  -1.040833
15        1.0        0.0   0.960769
16       -1.0        0.0  -1.040833
17        1.0        0.0   0.960769
18       -1.0        0.0  -1.040833
19        1.0        0.0   0.960769
20       -1.0        0.0  -1.040833
21        1.0        0.0   0.960769
22       -1.0        0.0  -1.040833
23

Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 7 seconds.
There were 107 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 136.36
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 3.3167     β0:0.564  β1:20.065  β2:1.119
Bayesian Variational Inference 3.3197     β0:0.435  β1:1.468  β2:1.390
Bayesian Linear Regression     3.3198     β0:0.449  β1:1.408  β2:1.378
Bayesian Robust Regression     3.3205     β0:0.439  β1:1.421  β2:1.404
Bayesian PCR                   3.3218     β0:0.706  β1:0.000  β2:0.811
Bayesian Ridge Regression      3.3262     β0:0.498  β1:1.431  β2:1.440
Bayesian ICR                   3.3319     β0:0.425  β1:0.000  β2:0.941

=== Running with parameters: {'high_corr': False, 'n_components': 3} ===
start
    feature_0  feature_1  feature_2
0        -1.0        0.0   0.960769
1         1.0        0.0   0.960769
2        -1.0        0.0  -1.040833
3         1.0        0.0   0.960769
4        -1.0        0.0  -1.040

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 7 seconds.
There were 240 divergences after tuning. Increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 134.63
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...
Failed to run Bayesian ICR: array must not contain infs or NaNs

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 3.2058     β0:0.654  β1:-4.377  β2:1.140
Bayesian Linear Regression     3.2087     β0:0.512  β1:1.400  β2:1.400
Bayesian Variational Inference 3.2087     β0:0.493  β1:1.487  β2:1.417
Bayesian Robust Regression     3.2089     β0:0.502  β1:1.416  β2:1.415
Bayesian PCR                   3.2109     β0:0.775  β1:-0.003  β2:0.840
Bayesian Ridge Regression      3.2168     β0:0.583  β1:1.380  β2:1.470
Bayesian ICR                   -1.0000    β0:nan  β1:nan  β2:nan

=== Running with parameters: {'high_corr': True, 'n_components': 2} ===
start
    feature_0  feature_1  feature_2
0   -1.549649   0.011881  -0.077415
1   -0.266309   0.010198  -0.244320
2    0.039225  -0.792

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 5 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
There were 56 divergences after tuning. Increase `target_accept` or reparameterize.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 153.55
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Ridge Regression      4.2867     β0:1.371  β1:0.117  β2:-0.660
Bayesian Lasso                 4.3237     β0:1.260  β1:0.386  β2:-0.559
Bayesian Linear Regression     4.3689     β0:1.308  β1:0.283  β2:-0.204
Bayesian Robust Regression     4.3872     β0:1.330  β1:0.363  β2:-0.157
Bayesian PCR                   4.4171     β0:0.520  β1:-0.662  β2:-0.302
Bayesian ICR                   4.4214     β0:0.494  β1:-0.625  β2:-0.282
Bayesian Variational Inference 4.4223     β0:1.246  β1:0.235  β2:0.043

=== Running with parameters: {'high_corr': True, 'n_components': 3} ===
start
    feature_0  feature_1  feature_2
0    0.253614  -0.270284  -0.708698
1   -0.303054   0.012421   0.329146
2    0.254646   0.123666   0.403828
3   -0.287135  -0.739111   0.626992
4   -0.451534  -0.306409  

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 6 seconds.
There were 114 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
There were 23 divergences after tuning. Increase `target_accept` or reparameterize.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 169.43
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian PCR                   6.3009     β0:0.332  β1:-0.113  β2:0.250
Bayesian ICR                   6.3017     β0:0.317  β1:-0.094  β2:0.255
Bayesian Lasso                 6.3209     β0:0.798  β1:0.251  β2:0.826
Bayesian Linear Regression     6.3728     β0:0.890  β1:0.524  β2:1.080
Bayesian Variational Inference 6.3783     β0:0.853  β1:0.544  β2:1.110
Bayesian Robust Regression     6.3787     β0:0.934  β1:0.521  β2:1.124
Bayesian Ridge Regression      6.4011     β0:0.871  β1:0.628  β2:1.196

=== Parameter Comparison Table ===
+--------------------------------+--------------+-------------+--------------+-------------+
| Model                          |   (2, False) |   (2, True) |   (3, False) |   (3, True) |
|--------------------------------+--------------+-------------+------

NameError: name 'results' is not defined

In [ ]:
plt.figure(figsize=(8, 6))

for model, rmse in rmse_results.items():
    if model != "ICR":
        plt.plot(sample_sizes, rmse, marker='o', label=model)
        
plt.xlabel("Sample Size")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Sample Size Increases")
plt.legend()
plt.grid()
plt.show()

### Number of parameters

In [ ]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

# Define parameter counts to test
param_counts = [3, 5, 10, 20, 50]  # Different numbers of parameters to test
model_names = [
    "Bayesian Linear Regression",
    "Bayesian Ridge Regression",
    "Bayesian Lasso", 
    "Bayesian Robust Regression",
    "Bayesian Variational Inference",
    "Bayesian PCR",
    "Bayesian ICR"
]

# Initialize results storage
results_data = []

# Fixed sample size for this analysis
n = 20

for p in param_counts:
    print(f"\n=== Running models with {p} parameters ===")
    
    # Generate true beta coefficients for current parameter count
    true_beta = np.logspace(0, 1, p, base=2)
    
    # Get results for current parameter count
    current_results = fourtenmodule.run_models_and_evaluate(n=n, p=p, true_beta=true_beta)
    
    # Verify results structure exists
    if 'results' not in current_results:
        print(f"Warning: No 'results' key found for p={p}")
        continue
        
    # Store results
    for result in current_results['results']:
        if result['Model'] in model_names:
            results_data.append({
                'Parameter Count': p,
                'Model': result['Model'],
                'RMSE': result['RMSE'],
                'Beta Error': np.mean(np.abs(np.array(result.get('Beta', [])) - true_beta)) 
                # Adding beta error metric
            })

# Convert to DataFrame
results_df = pd.DataFrame(results_data)

# Pivot for plotting if we have data
if not results_df.empty:
    # Plot RMSE as function of parameter count
    plt.figure(figsize=(12, 5))
    
    plt.subplot(1, 2, 1)
    rmse_data = results_df.pivot(index='Parameter Count', columns='Model', values='RMSE')
    for model in model_names:
        if model in rmse_data.columns:
            plt.plot(rmse_data.index, rmse_data[model], marker='o', label=model)
    plt.xlabel("Number of Parameters")
    plt.ylabel("RMSE")
    plt.title("Prediction Error vs Parameter Count")
    plt.legend()
    plt.grid(True)
    
    # Plot Beta Error as function of parameter count
    plt.subplot(1, 2, 2)
    beta_error_data = results_df.pivot(index='Parameter Count', columns='Model', values='Beta Error')
    for model in model_names:
        if model in beta_error_data.columns:
            plt.plot(beta_error_data.index, beta_error_data[model], marker='o', label=model)
    plt.xlabel("Number of Parameters")
    plt.ylabel("Mean Absolute Beta Error")
    plt.title("Parameter Estimation Error vs Parameter Count")
    plt.grid(True)
    
    plt.tight_layout()
    plt.show()
else:
    print("No valid results to plot")

# Show complete results
print("\nComplete Results:")
print(results_df)

# Additional analysis
if not results_df.empty:
    print("\nPerformance Summary:")
    summary_df = results_df.groupby('Model').agg({
        'RMSE': ['mean', 'std'],
        'Beta Error': ['mean', 'std']
    })
    print(summary_df)

Initializing NUTS using jitter+adapt_diag...



=== Running models with 3 parameters ===
heeleloeo

=== Running with parameters: {'high_corr': False, 'n_components': 2} ===
start
    feature_0  feature_1  feature_2
0        -1.0        0.0   0.904534
1         1.0        0.0   0.904534
2        -1.0        0.0  -1.105542
3         1.0        0.0   0.904534
4        -1.0        0.0  -1.105542
5         1.0        0.0   0.904534
6        -1.0        0.0  -1.105542
7         1.0        0.0   0.904534
8        -1.0        0.0  -1.105542
9         1.0        0.0   0.904534
10       -1.0        0.0  -1.105542
11        1.0        0.0   0.904534
12       -1.0        0.0  -1.105542
13        1.0        0.0   0.904534
14       -1.0        0.0  -1.105542
15        1.0        0.0   0.904534
16       -1.0        0.0  -1.105542
17        1.0        0.0   0.904534
18       -1.0        0.0  -1.105542
19        1.0        0.0   0.904534 0     5.190469
1     3.594513
2     0.194228
3     3.888922
4     3.408981
5     2.808905
6     1.308407
7     3

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 7 seconds.
There were 353 divergences after tuning. Increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 54.536
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 2.9968     β0:-0.482  β1:1.487  β2:1.750
Bayesian ICR                   3.0300     β0:-0.417  β1:0.000  β2:1.234
Bayesian Linear Regression     3.0307     β0:0.182  β1:1.411  β2:1.431
Bayesian Variational Inference 3.0309     β0:0.237  β1:1.453  β2:1.355
Bayesian Robust Regression     3.0345     β0:0.210  β1:1.398  β2:1.430
Bayesian PCR                   3.0418     β0:0.282  β1:0.000  β2:0.692
Bayesian Ridge Regression      3.0546     β0:0.228  β1:1.423  β2:1.562

=== Running with parameters: {'high_corr': False, 'n_components': 3} ===
start
    feature_0  feature_1  feature_2
0        -1.0        0.0   0.904534
1         1.0        0.0   0.904534
2        -1.0        0.0  -1.105542
3         1.0        0.0   0.904534
4        -1.0        0.0  -1.10

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 7 seconds.
There were 285 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 55.416
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...
Failed to run Bayesian ICR: array must not contain infs or NaNs

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 3.2303     β0:0.458  β1:-3.646  β2:1.507
Bayesian Variational Inference 3.2384     β0:0.590  β1:1.364  β2:1.616
Bayesian Linear Regression     3.2390     β0:0.612  β1:1.423  β2:1.603
Bayesian Robust Regression     3.2394     β0:0.605  β1:1.424  β2:1.614
Bayesian Ridge Regression      3.2590     β0:0.700  β1:1.397  β2:1.713
Bayesian PCR                   3.2675     β0:0.675  β1:0.007  β2:0.830
Bayesian ICR                   -1.0000    β0:nan  β1:nan  β2:nan

=== Running with parameters: {'high_corr': True, 'n_components': 2} ===
start
    feature_0  feature_1  feature_2
0   -2.102438   0.371612  -0.881126
1    0.480266   0.035918   0.132586
2   -0.139481  -0.4476

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

In [ ]:
# Define sample sizes to test
p_numbers = [5, 20, 30, 60, 100]
n=20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR" : [], "ICR": []}

for p in p_numbers:
    print(f"Running models for this many parameters: {p}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(p = p)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of sample size
plt.figure(figsize=(8, 6))
for model, rmse in rmse_results.items():
    plt.plot(p_numbers, rmse, marker='o', label=model)
plt.xlabel("Number of Predictors")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Number of Predictors Increases")
plt.legend()
plt.grid()
plt.show()

Running models for this many parameters: 5

Running Bayesian Linear Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 4 seconds.
There were 1 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian Linear Regression: name 'mean_squared_error' is not defined

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 11 seconds.
There were 69 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian Ridge Regression: name 'mean_squared_error' is not defined

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 10 seconds.
There were 192 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian Lasso: name 'mean_squared_error' is not defined

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 5 seconds.


Failed to run Bayesian Robust Regression: name 'mean_squared_error' is not defined

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 37.785
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian Variational Inference: name 'mean_squared_error' is not defined

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian PCR: name 'mean_squared_error' is not defined

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian ICR: name 'mean_squared_error' is not defined

=== Final Results ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Linear Regression     -1.0000    N/A
Bayesian Ridge Regression      -1.0000    N/A
Bayesian Lasso                 -1.0000    N/A
Bayesian Robust Regression     -1.0000    N/A
Bayesian Variational Inference -1.0000    N/A
Bayesian PCR                   -1.0000    N/A
Bayesian ICR                   -1.0000    N/A
Running models for this many parameters: 20

Running Bayesian Linear Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 8 seconds.
There were 44 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian Linear Regression: name 'mean_squared_error' is not defined

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 90 seconds.
There were 6 divergences after tuning. Increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian Ridge Regression: name 'mean_squared_error' is not defined

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 442 draw iterations (4_000 + 1_768 draws total) took 29 seconds.
There were 41 divergences after tuning. Increase `target_accept` or reparameterize.
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+a

Failed to run Bayesian Lasso: name 'mean_squared_error' is not defined

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 9 seconds.


Failed to run Bayesian Robust Regression: name 'mean_squared_error' is not defined

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 55.916
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian Variational Inference: name 'mean_squared_error' is not defined

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Failed to run Bayesian PCR: name 'mean_squared_error' is not defined

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

### Using highly sparse data

In [ ]:
# Define sample sizes to test
p_numbers = [5, 20, 30, 60, 100]
n=20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR" : [], "ICR": []}

for p in p_numbers:
    print(f"Running models for this many parameters: {p}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(p = p, high_corr = True)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of sample size
plt.figure(figsize=(8, 6))
for model, rmse in rmse_results.items():
    plt.plot(p_numbers, rmse, marker='o', label=model)
plt.xlabel("Number of Predictors")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Number of Predictors Increases With Sparse True Signals")
plt.legend()
plt.grid()
plt.show()

### Magnitudes

#### Beta 0

In [ ]:
# Define different values where only one coefficient increases at a time
p=5
# Fixed 5-element true_beta array
true_beta = np.logspace(0, 1, p, base=2)

beta_variations = [[2**i, *true_beta[1:]] for i in range(5)]


n = 20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR": [], "ICR": []}

for true_beta in beta_variations:
    print(f"Running models with true_beta: {true_beta}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(n = n, true_beta = true_beta)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of increasing individual beta values
plt.figure(figsize=(8, 6))
beta_labels = [
    "1", "2", "4", "8", "16"
]
for model, rmse in rmse_results.items():
    plt.plot(beta_labels, rmse, marker='o', label=model)
plt.xlabel("True Beta 0 Variations")
plt.ylabel("RMSE of Estimated Beta 0")
plt.title("Model Performance as Beta 0 Increases")
plt.xticks(rotation=45)
plt.legend()
plt.grid()
plt.show()


#### Beta 1

In [ ]:
p=5
# Fixed 5-element true_beta array
true_beta = np.logspace(0, 1, p, base=2)

# Define different values where only one coefficient increases at a time
beta_variations = [ [true_beta[0], 2**i, *true_beta[2:]] for i in range(5)]


n = 20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR": [], "ICR": []}

for true_beta in beta_variations:
    print(f"Running models with true_beta: {true_beta}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(n = n, true_beta = true_beta)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of increasing individual beta values
plt.figure(figsize=(8, 6))
beta_labels = [
    "1", "1.41", "4", "8", "16"
]
for model, rmse in rmse_results.items():
    plt.plot(beta_labels, rmse, marker='o', label=model)
plt.xlabel("True Beta 1 Variations")
plt.ylabel("RMSE of Estimated Beta 1")
plt.title("Model Performance as Beta 1 Increases")
plt.xticks(rotation=45)
plt.legend()
plt.grid()
plt.show()


# Real Data

## Housing Data

In [3]:
def load_soccer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/refs/heads/main/housing.csv"
        return pd.read_csv(url, header=None, sep=r"\s+")
    except:
        try:
            return pd.read_csv("housing.csv", header=None, sep=r"\s+")
        except:
            raise Exception("Could not load data")

housing = load_soccer_data()
# Load dataset
housing.columns = ["CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE", "DIS", "RAD", "TAX", "PTRATIO", "B", "LSTAT", "MEDV"]





results = fourtenmodule.run_and_plot_models(housing, "MEDV", n_components=2)




Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.


start
         CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
0   -0.419782  0.284830 -1.287909 -0.272599 -0.144217  0.413672 -0.120013   
1   -0.417339 -0.487722 -0.593381 -0.272599 -0.740262  0.194274  0.367166   
2   -0.417342 -0.487722 -0.593381 -0.272599 -0.740262  1.282714 -0.265812   
3   -0.416750 -0.487722 -1.306878 -0.272599 -0.835284  1.016303 -0.809889   
4   -0.412482 -0.487722 -1.306878 -0.272599 -0.835284  1.228577 -0.511180   
..        ...       ...       ...       ...       ...       ...       ...   
501 -0.413229 -0.487722  0.115738 -0.272599  0.158124  0.439316  0.018673   
502 -0.415249 -0.487722  0.115738 -0.272599  0.158124 -0.234548  0.288933   
503 -0.413447 -0.487722  0.115738 -0.272599  0.158124  0.984960  0.797449   
504 -0.407764 -0.487722  0.115738 -0.272599  0.158124  0.725672  0.736996   
505 -0.415000 -0.487722  0.115738 -0.272599  0.158124 -0.362767  0.434732   

          DIS       RAD       TAX   PTRATIO         B     LSTAT  
0  

Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 2.71s

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.68s

Testing Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.73s

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 0.96s

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 2,416.5
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.25s

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.65s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.48s

=== Running full analysis on passing models ===

Running Bayesian Linear Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 3 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 9 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 4 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 4 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 2,418.4
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 1 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 23.0218    β0:-0.744  β1:0.783  β2:-0.207  β3:0.680  β4:-1.433  β5:2.787  β6:-0.077  β7:-2.463  β8:1.484  β9:-1.126  β10:-1.831  β11:0.776  β12:-3.554
Bayesian Ridge Regression      23.1011    β0:-0.600  β1:0.488  β2:-0.476  β3:0.688  β4:-0.671  β5:2.491  β6:-0.205  β7:-1.126  β8:0.312  β9:-0.504  β10:-1.437  β11:0.683  β12:-2.468
Bayesian Variational Inference 23.1596    β0:-0.641  β1:0.529  β2:-0.626  β3:0.802  β4:-0.542  β5:2.082  β6:-0.263  β7:-0.845  β8:0.110  β9:-0.428  β10:-1.471  β11:0.699  β12:-1.962
Bayesian Linear Regression     23.2183    β0:-0.504  β1:0.413  β2:-0.495  β3:0.575  β4:-0.450  β5:1.965  β6:-0.259  β7:-0.487  β8:-0.068  β9:-0.455  β10:-1.136  β11:0.554  β12:-1.812
Bayesian Robust Regression     23.2255    β0:-0.508  β1:0.403

## Soccer Players' Market Value
Many redundant and multicollinear variables, also too many features, expect most algorithms to time out and ICR or PCR to perform better

In [5]:
# Load dataset (with SSL fix)
def load_soccer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/main/transfermarkt201718.csv"
        return pd.read_csv(url, sep=";")
    except:
        try:
            return pd.read_csv("transfermarkt201718.csv", sep=";")
        except:
            raise Exception("Could not load data")

soccer = load_soccer_data()


print(len(soccer))
# Preprocess data and select top 100 features
#X, y = fourtenmodule.preprocess_data(soccer, "value")


# Now run models
results = fourtenmodule.run_and_plot_models(soccer, "value")



2232
start

Selected top 399 features from 201707 total features
Top 10 features by variance:
- dribbles_completedm_nan
- through_ballsm_nan
- sca_passes_livem_nan
- passes_pressurem_nan
- gca_og_form_nan
- passes_blockedm_nan
- gca_fouledm_nan
- passes_switchesm_nan
- crossesm_nan
- sca_passes_deadm_nan


Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


      dribbles_completedm_nan  through_ballsm_nan  sca_passes_livem_nan  \
0                         0.0                 0.0                   0.0   
1                         0.0                 0.0                   0.0   
2                         0.0                 0.0                   0.0   
3                         0.0                 0.0                   0.0   
4                         0.0                 0.0                   0.0   
...                       ...                 ...                   ...   
2227                      0.0                 0.0                   0.0   
2228                      0.0                 0.0                   0.0   
2229                      0.0                 0.0                   0.0   
2230                      1.0                 1.0                   1.0   
2231                      1.0                 1.0                   1.0   

      passes_pressurem_nan  gca_og_form_nan  passes_blockedm_nan  \
0                      0.0     

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Failed: 

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

  → Failed: 

Testing Bayesian Lasso...


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:681: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 46048632352672.0, tolerance: 40333765302331.92
  model = cd_fast.enet_coordinate_descent_gram(
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:681: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 43684367241072.0, tolerance: 40333765302331.92
  model = cd_fast.enet_coordinate_descent_gram(
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:681: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 42844392907248.0, tolerance: 40333765302331.92
  model = cd_fast.enet_coordinate_

Output()

Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Failed: 

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

  → Failed: 

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 1.2344e+15
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 4.27s

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.41s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 0.61s

=== Models that didn't pass timing tests ===
Bayesian Linear Regression: 
Bayesian Ridge Regression: 
Bayesian Lasso: 
Bayesian Robust Regression: 

=== Running full analysis on passing models ===

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 1.1033e+15
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Variational Inference 19246697.3205 β0:2.472  β1:2.516  β2:2.476  β3:2.514  β4:2.467  β5:2.504  β6:2.527  β7:2.499  β8:2.473  β9:2.505  β10:2.439  β11:2.464  β12:2.487  β13:2.478  β14:2.451  β15:2.502  β16:2.484  β17:2.502  β18:2.486  β19:2.462  β20:2.470  β21:2.488  β22:2.496  β23:2.499  β24:2.449  β25:2.488  β26:2.476  β27:2.475  β28:2.474  β29:2.491  β30:2.498  β31:2.464  β32:2.482  β33:2.508  β34:2.484  β35:2.508  β36:2.472  β37:2.489  β38:2.473  β39:2.475  β40:2.470  β41:2.500  β42:2.502  β43:2.533  β44:2.485  β45:2.488  β46:2.444  β47:2.514  β48:2.466  β49:2.520  β50:2.489  β51:2.464  β52:2.489  β53:2.486  β54:2.514  β55:2.512  β56:2.511  β57:2.493  β58:2.463  β59:2.458  β60:2.465  β61:2.483  β62:2.509  β63:2.511  β64:2.478  β65:2.508  β66:2.464  β67:2.526  β68:2.5

### Cancer Survival Data
Highly collinear

In [2]:
def load_cancer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/refs/heads/main/METABRIC_RNA_Mutation.csv"
        return pd.read_csv(url, sep=";")
    except:
        try:
            return pd.read_csv("METABRIC_RNA_Mutation.csv")
        except:
            raise Exception("Could not load data")

cancer = load_cancer_data()
# Load dataset

# Define X (predictors) and y (target variable)

X, y = fourtenmodule.preprocess_data(cancer, "overall_survival")

# Normalize X
X = (X - X.mean(axis=0)) / X.std(axis=0)



results = fourtenmodule.run_and_plot_models(X, y, n_components=2)




/var/folders/x9/3zl22m596tl6d44cbs62z6h00000gn/T/ipykernel_23385/1295189474.py:7: DtypeWarning: Columns (678,688,690,692) have mixed types. Specify dtype option on import or set low_memory=False.
  return pd.read_csv("METABRIC_RNA_Mutation.csv")
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.



Selected top 692 features from 8658 total features
Top 10 features by variance:
- pbrm1_mut_0
- er_status_measured_by_ihc_nan
- l1cam_mut_0
- tumor_size_14.0
- ncoa3_mut_0
- bap1_mut_0
- kdm3a_mut_0
- sbno1_mut_0
- lifr_mut_0
- aff2_mut_0

=== Running timing tests (max 5s per model) ===

Testing Bayesian Linear Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 8 seconds.
There were 43 divergences after tuning. Increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Failed: took 12.39s (timeout)

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 31 seconds.
There were 59 divergences after tuning. Increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.


  → Failed: took 33.14s (timeout)

Testing Bayesian Lasso...


Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
There were 55 divergences after tuning. Increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.95s

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 9 seconds.
There were 37 divergences after tuning. Increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.


  → Failed: took 11.22s (timeout)

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 7,626
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Failed: took 7.58s (timeout)

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
There were 2 divergences after tuning. Increase `target_accept` or reparameterize.
The number of samples is too small to check convergence reliably.
Only 15 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.56s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 15 tune and 15 draw iterations (60 + 60 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 0.59s

=== Models that didn't pass timing tests ===
Bayesian Linear Regression: Too slow (12.39s > 5s)
Bayesian Ridge Regression: Too slow (33.14s > 5s)
Bayesian Robust Regression: Too slow (11.22s > 5s)
Bayesian Variational Inference: Too slow (7.58s > 5s)

=== Running full analysis on passing models ===

Running Bayesian Lasso...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 734 draw iterations (4_000 + 2_936 draws total) took 665 seconds.
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 0.4272     β0:0.002  β1:-1.750  β2:0.001  β3:0.002  β4:-0.001  β5:-0.001  β6:0.006  β7:0.001  β8:-0.006  β9:0.001  β10:-0.000  β11:-0.000  β12:-0.007  β13:-0.006  β14:0.003  β15:-0.002  β16:-0.004  β17:-0.007  β18:-0.004  β19:-0.001  β20:-0.004  β21:0.006  β22:0.003  β23:0.001  β24:-0.002  β25:0.000  β26:0.000  β27:0.002  β28:-0.004  β29:-0.003  β30:0.002  β31:0.001  β32:0.008  β33:-0.001  β34:-0.006  β35:0.003  β36:0.007  β37:-0.001  β38:0.001  β39:0.001  β40:-0.001  β41:-0.007  β42:-0.003  β43:0.006  β44:-0.001  β45:-0.003  β46:-0.003  β47:-0.001  β48:0.000  β49:-0.002  β50:-0.007  β51:-0.002  β52:0.001  β53:-0.000  β54:-0.384  β55:0.005  β56:-0.011  β57:0.002  β58:0.002  β59:0.004  β60:0.001  β61:-0.005  β62:-0.007  β63:0.004  β64:0.004  β65:-0.0